### PR TITLE
Feat/crypto setup

### PR DESCRIPTION
**Pull request by magsteen**
Added functions for easily generating keypair(s) and encrypt/decrypt byte vectors. 

Added one unit test for encryption/decryption.
Dropped test for keygeneration as it is mostly handled by an external create and our function will merely return other types in a wrapper. 

What do you guys think? Is the encryption valid? Anything more we should add to crypto.rs?

image

### CODE REVIEW

File: core/src/crypto.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+fn create_key() -> KeyPair {
+    let secret = EphemeralSecret::new(OsRng);
+    let public_key = PublicKey::from(&secret);
+    KeyPair { secret, public_key }
+}
+
+fn create_keys(n: u8) -> Vec<KeyPair> {
+    let mut keys = Vec::new();
```
Comment (ngorke) [inline core/src/crypto.rs]: You can use Vec::with_capacity here to pre-allocate.

File: core/src/crypto.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_create_key() {}
+
+    #[test]
+    fn test_create_keys() {}
```
Comment (ngorke) [inline core/src/crypto.rs]: You can remove the functions if you aren't going to have the tests

File: core/src/crypto.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    #[test]
+    fn test_xor() {
+        let mut pair = create_key();
+
+        let msg = "Hello".to_owned().into_bytes();
+
+        let encrypted_msg = pair.public_key.xor(&msg);
+        let unencrypted_msg = pair.public_key.xor(&encrypted_msg);
+
+        assert_eq!(msg, unencrypted_msg);
```
Comment (ngorke) [inline core/src/crypto.rs]: You should rather test for an expected output, not if the two invocations cancel each other.

File: core/src/crypto.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                for i in 0..data.len() {
+                    res.push(data[i] ^ key_bytes[i % key_len])
+                }
+                return res
+            }
+        })*
+    }
+}
+
+impl_key!(SharedSecret, PublicKey);
```
Comment (ngorke) [inline core/src/crypto.rs]: Why do we need xor?

File: core/src/crypto.rs [hunk 5]
```diff
@@ -0,0 +1,66 @@
+use rand_core::OsRng;
+use x25519_dalek::{EphemeralSecret, PublicKey, SharedSecret};
+
+struct KeyPair {
+    secret: EphemeralSecret,
+    public_key: PublicKey,
+}
```
Comment (ngorke) [inline core/src/crypto.rs]: This is not a key pair.

**Review by ngorke — CHANGES_REQUESTED** [review]

File: core/src/crypto.rs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+fn create_key() -> KeyPair {
+    let secret = EphemeralSecret::new(OsRng);
+    let public_key = PublicKey::from(&secret);
+    KeyPair { secret, public_key }
+}
```
Comment (ngorke) [inline core/src/crypto.rs]: Also, this should be a new() method in the KeyPair struct.

File: core/src/crypto.rs (hunk 2 above)
Comment (magsteen) [reply core/src/crypto.rs]: Yes, I agree. Was wondering if we should test this somehow?

File: core/src/crypto.rs (hunk 3 above)
Comment (magsteen) [reply core/src/crypto.rs]: Yes :point_right:

File: core/src/crypto.rs (hunk 1 above)
Comment (magsteen) [reply core/src/crypto.rs]: Good point!

File: core/src/crypto.rs (hunk 5 above)
Comment (magsteen) [reply core/src/crypto.rs]: What do you think we should call it?

File: core/src/crypto.rs (hunk 4 above)
Comment (magsteen) [reply core/src/crypto.rs]: Well I do realize now that it is useless for the public key, but still useful for encrypting/decrypting with a symmetric key.

File: core/src/crypto.rs (hunk 5 above)
Comment (ngorke) [reply core/src/crypto.rs]: Secret :feelsgood:

File: core/src/crypto.rs (hunk 4 above)
Comment (ngorke) [reply core/src/crypto.rs]: We cannot simply use XOR for encryption/decryption, as it is not safe to use it for symmetric encryption when we reuse the encryption key.

File: core/src/crypto.rs (hunk 2 above)
Comment (ngorke) [reply core/src/crypto.rs]: Yes, but I don't think it will provide much value.

**Comment by magsteen** [comment]
Refactored and added a AES library we can use. We only have to generate ciphers and use the inbuilt encrypt/decrypt functions. The library provides a good variety of functions i think. :)

File: core/src/lib.rs [hunk 7]
```diff
@@ -1 +1,2 @@
-// lib code
\ No newline at end of file
+// lib code
```
Comment (ngorke) [inline core/src/lib.rs]: plz no

**Review by ngorke — CHANGES_REQUESTED** [review]

File: core/Cargo.toml [hunk 8]
```diff
@@ -9,3 +9,6 @@ Library for using the onion protocol to securely and anonymously connect to serv
 repository = "https://github.com/tomrsae/ronion"
 
 [dependencies]
+x25519-dalek = "2.0.0-pre.1"
```
Comment (ngorke) [inline core/Cargo.toml]: 🍣 -hi

File: core/src/crypto.rs [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+            secret: EphemeralSecret::new(OsRng),
+        }
+    }
+
+    pub fn create_secrets(n: usize) -> Vec<Secret> {
+        let mut secrets = Vec::with_capacity(n);
+        for _ in 0..n {
+            secrets.push(Secret::new())
+        }
+        secrets
```
Comment (ngorke) [inline core/src/crypto.rs]: You could also do
`
(0..n).iter().map(|_| Secret::new()).collect()
`

File: core/src/crypto.rs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        PublicKey::from(&self.secret)
+    }
+
+    pub fn gen_shared_key(self, pub_key: &PublicKey) -> SharedSecret {
+        self.secret.diffie_hellman(pub_key)
+    }
+
+    pub fn gen_cipher(&mut self) -> Aes256 {
+        let key = self.gen_pub_key();
+        match Aes256::new_from_slice(key.as_bytes()) {
```
Comment (ngorke) [inline core/src/crypto.rs]: You could also just call .expect on the result

**Review by ngorke — CHANGES_REQUESTED** [review]
🍣 -hi

**Comment by tomrsae** [comment]
Where are the tests? I can see the screenshot of it, but they're never ran, nor can I find the code for it 🤔

**Review by ngorke — APPROVED** [review]
Nice r🧅!

**Pull request merged**